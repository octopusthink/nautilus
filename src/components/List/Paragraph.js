import React from 'react';

export const ListParagraph = (props) => {
  return <React.Fragment {...props} />;
};

export default ListParagraph;
