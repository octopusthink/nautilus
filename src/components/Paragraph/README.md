A `Paragraph` is a self-contained unit of one or more coherent sentences related to a single topic or idea.

```jsx
<Paragraph size="large">
	Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut bibendum hendrerit
	augue gravida hendrerit. Vestibulum ornare eget lacus cursus rhoncus. Cras nec
	lectus in sem viverra volutpat.
</Paragraph>
```

## Usage

Use a `Paragraph` for any body copy comprising of one or more full sentences.

Don't use a `Paragraph` for:
- A single text link or call to action—use a `Button` instead.
- A heading or subheading—use a `Heading`, `Subtitle`, or `Page Title`.
- A list of related, equally weighted items—use a `List` instead.

## Appearance

- Line lengths should be between 55-70 characters.
- Text colour should contrast with the background.
- Use the large and small variants sparingly when hierarchy is required. Most `Paragraphs` should use the default size.

Use a **large paragraph** when you want to emphasise an important passage of text.

```jsx
<Paragraph>
	Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut bibendum hendrerit
	augue gravida hendrerit. Vestibulum ornare eget lacus cursus rhoncus. Cras nec
	lectus in sem viverra volutpat.
</Paragraph>
```

Use a **small paragraph** for fine print, or details that are secondary to the main content.

```jsx
<Paragraph size="small">
	Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut bibendum hendrerit
	augue gravida hendrerit. Vestibulum ornare eget lacus cursus rhoncus. Cras nec
	lectus in sem viverra volutpat.
</Paragraph>
```


## Voice & Tone

```jsx
<Paragraph color="accent">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="medium">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="light">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="danger">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="success">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
```
