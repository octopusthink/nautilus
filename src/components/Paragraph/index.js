import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';

import { textStyles } from 'themes/mixins';

export const Paragraph = ({
  children,
  large,
  small,
  inverse,
  dark,
  light,
  ...otherProps
}) => {
  return <p {...otherProps}>{children}</p>;
};

Paragraph.defaultProps = {
  children: undefined,
  large: false,
  small: false,
  inverse: false,
  dark: false,
  light: false,
};

Paragraph.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Increase the visual prominence of the paragraph. */
  large: PropTypes.bool,
  /** Decrease the visual prominence of the paragraph. */
  small: PropTypes.bool,
  /** Inverse text colour. Used for dark backgrounds. */
  inverse: PropTypes.bool,
  /** Darken text colour. */
  dark: PropTypes.bool,
  /** Lighten text colour. */
  light: PropTypes.bool,
};

const StyledParagraph = styled(Paragraph)(textStyles);

export default StyledParagraph;
