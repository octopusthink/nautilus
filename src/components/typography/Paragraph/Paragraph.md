## Available Sizes
### Large Paragraph
```js
<Paragraph size="large">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut bibendum hendrerit augue gravida hendrerit. Vestibulum ornare eget lacus cursus rhoncus. Cras nec lectus in sem viverra volutpat. Fusce placerat nunc eu libero egestas, sit amet condimentum turpis ornare. Sed hendrerit metus elit, ac sollicitudin enim vulputate et. Ut fermentum venenatis auctor. Mauris augue augue, hendrerit quis risus eget, imperdiet feugiat odio. Nullam porta sollicitudin ipsum eu varius. Nam hendrerit eleifend sagittis. Maecenas in sodales libero, nec ultrices ante. Ut et sem nibh. Vivamus fringilla pharetra justo aliquet pharetra.</Paragraph>
```

### Medium Paragraph
```js
<Paragraph>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut bibendum hendrerit augue gravida hendrerit. Vestibulum ornare eget lacus cursus rhoncus. Cras nec lectus in sem viverra volutpat. Fusce placerat nunc eu libero egestas, sit amet condimentum turpis ornare. Sed hendrerit metus elit, ac sollicitudin enim vulputate et. Ut fermentum venenatis auctor. Mauris augue augue, hendrerit quis risus eget, imperdiet feugiat odio. Nullam porta sollicitudin ipsum eu varius. Nam hendrerit eleifend sagittis. Maecenas in sodales libero, nec ultrices ante. Ut et sem nibh. Vivamus fringilla pharetra justo aliquet pharetra.</Paragraph>
```

### Small (Fine Print) Paragraph
```js
<Paragraph size="small">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut bibendum hendrerit augue gravida hendrerit. Vestibulum ornare eget lacus cursus rhoncus. Cras nec lectus in sem viverra volutpat. Fusce placerat nunc eu libero egestas, sit amet condimentum turpis ornare. Sed hendrerit metus elit, ac sollicitudin enim vulputate et. Ut fermentum venenatis auctor. Mauris augue augue, hendrerit quis risus eget, imperdiet feugiat odio. Nullam porta sollicitudin ipsum eu varius. Nam hendrerit eleifend sagittis. Maecenas in sodales libero, nec ultrices ante. Ut et sem nibh. Vivamus fringilla pharetra justo aliquet pharetra.</Paragraph>
```

## Colors
### Paragraph with custom color
```js
<Paragraph color="accent">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="medium">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="light">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="danger">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
<Paragraph color="success">Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
```

## Margins
### Paragraph with no margin
```js
<Paragraph noMargin>Aenean auctor justo nisl, in ullamcorper turpis pharetra mattis. Duis eget vulputate est. Nullam a velit pharetra, hendrerit magna vitae, accumsan magna.</Paragraph>
```
