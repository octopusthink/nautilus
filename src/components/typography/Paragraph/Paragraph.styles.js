import styled from 'styled-components';

import typography from '../../../base/typography/typography';
import { semanticColors, typographicColors } from '../../../base/color/colorVariables';

export const paragraphSizes = {
  small: typography.finePrint,
  medium: typography.body,
  large: typography.bodyLarge,
};

export const paragraphColors = {
  light: typographicColors.typeLight,
  medium: typographicColors.typeMedium,
  dark: typographicColors.typeDark,
  accent: typographicColors.typeAccent,
  success: semanticColors.colorSuccess,
  danger: semanticColors.colorDanger,
};

export const StyledParagraph = styled.p`
  color: ${p => paragraphColors[p.color]};
  ${p => paragraphSizes[p.size]};
  ${p => p.noMargin && 'margin: 0'};
`;
