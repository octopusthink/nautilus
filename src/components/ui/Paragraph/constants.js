export const ParagraphClassName = 'Nautilus-Paragraph';
