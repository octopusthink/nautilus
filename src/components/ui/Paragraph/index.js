import { css } from '@emotion/core';
import classnames from 'classnames';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import React from 'react';

import { bodyStyles } from 'styles';
import { useTheme } from 'themes';

export const ComponentClassName = 'Nautilus-Paragraph';

export const qualityControl = (props) => {
  const { dark, large, light, small } = props;

  invariant(
    [dark, light].filter((prop) => prop).length <= 1,
    'Paragraph cannot have both `dark` and `light` props set.',
  );

  invariant(
    [large, small].filter((prop) => prop).length <= 1,
    'Paragraph cannot have both `large` and `small` props set.',
  );
};

export const Paragraph = (props) => {
  const {
    children,
    className,
    dark,
    inverse,
    large,
    light,
    small,
    unstyled,
    ...otherProps
  } = props;

  qualityControl(props);

  const theme = useTheme();

  return (
    <p
      className={classnames(ComponentClassName, className)}
      css={
        unstyled
          ? undefined
          : css`
              ${bodyStyles({
                large,
                small,
                inverse,
                dark,
                light,
                theme,
              })}
            `
      }
      {...otherProps}
    >
      {children}
    </p>
  );
};

Paragraph.defaultProps = {
  children: undefined,
  className: undefined,
  large: false,
  small: false,
  inverse: false,
  dark: false,
  light: false,
  unstyled: false,
};

Paragraph.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  className: PropTypes.string,
  /** Increase the visual prominence of the paragraph. */
  large: PropTypes.bool,
  /** Decrease the visual prominence of the paragraph. */
  small: PropTypes.bool,
  /** Inverse text colour. Used for dark backgrounds. */
  inverse: PropTypes.bool,
  /** Darken text colour. */
  dark: PropTypes.bool,
  /** Lighten text colour. */
  light: PropTypes.bool,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = Paragraph;

export default Paragraph;
