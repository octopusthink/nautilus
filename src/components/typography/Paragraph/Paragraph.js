import PropTypes from 'prop-types';
import React from 'react';

import { StyledParagraph, paragraphSizes, paragraphColors } from './Paragraph.styles';

export const Paragraph = ({ children, color, noMargin, size, ...others }) => (
  <StyledParagraph
    color={color}
    noMargin={noMargin}
    size={size}
    {...others}
  >
    {children}
  </StyledParagraph>
);

export const ParagraphSizes = Object.keys(paragraphSizes);
export const ParagraphColors = Object.keys(paragraphColors);

Paragraph.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Color of the Paragraph. Options are 'dark', 'medium', 'light', 'accent', 'success', and 'danger' */
  color: PropTypes.oneOf(ParagraphColors),
  /** If set, removes margin on Paragraph instance */
  noMargin: PropTypes.bool,
  /** Size of the Paragraph. Options are 'small', 'medium', 'large' */
  size: PropTypes.oneOf(ParagraphSizes),
};

Paragraph.defaultProps = {
  color: 'dark',
  size: 'medium',
};
