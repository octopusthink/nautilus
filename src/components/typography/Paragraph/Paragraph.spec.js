import React from 'react';

import { Paragraph, ParagraphSizes, ParagraphColors } from './Paragraph';

describe('Paragraph Component', () => {
  context('when nothing specified', () => {
    it('should render correctly', () => {
      const subject = shallow(<Paragraph>Hello</Paragraph>);
      expect(subject).toMatchSnapshot();
    });
  });

  ParagraphSizes.forEach((size) => {
    context(`when @size specified as ${size}`, () => {
      it(`renders a Paragraph size of ${size}`, () => {
        const subject = mount(<Paragraph size={size}>Hello</Paragraph>);
        expect(subject.prop('size')).toBe(size);
      });
    });
  });

  ParagraphColors.forEach((color) => {
    context(`when @color specified as ${color}`, () => {
      it(`renders a Paragraph color of ${color}`, () => {
        const subject = mount(<Paragraph color={color}>Hello</Paragraph>);
        expect(subject.prop('color')).toBe(color);
      });
    });
  });
});
