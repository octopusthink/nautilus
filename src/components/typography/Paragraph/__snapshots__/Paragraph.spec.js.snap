// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph Component when nothing specified should render correctly 1`] = `
<ForwardRef
  color="dark"
  size="medium"
>
  Hello
</ForwardRef>
`;
